emma-language/src/main/scala/org/emmalanguage/compiler/lang/FunSupport.scala

```diff
@@ -0,0 +1,478 @@
+                  else substBody match {
+                    case Block(stats, expr) => Block(prefxSeq ++ stats: _*)(expr)
+                    case expr => Block(prefxSeq: _*)(expr)
+                  }
+
+                val anyEq = u.typeOf[Any].member(TermName("==")).asMethod
+                val fixEqBody = BottomUp.transform({
+                  case DefCall(Some(lhs), `anyEq`, Seq(), Seq(rhs)) =>
+                    val specializedEq = Type.of(lhs).member(anyEq.name).asTerm
+                    DefCall(Some(lhs))(specializedEq)(Seq(rhs))
+                })(prefxBody).tree
+
+                Owner.at(owner)(fixEqBody)
+            })(snippet.tree).tree
+
+          result
+      }
+
+      CG.Snippet(result)
+    }
+
+    private lazy val calleAttrs: u.Tree => (Set[u.TermSymbol], Map[u.TermSymbol, Int]) =
+      BottomUp.withParUses.withBindDefs.traverseAny andThen {
+        case Attr.syn(_, bndDefs :: parUses :: _) => (bndDefs.keySet, parUses)
+      }
+
+    lazy val callGraph: u.Tree => CG = tree => {
+      @tailrec
+      def discover(cg: CG, frontier: Set[CG.TreeRef] = Set.empty): CG =
+        frontier.headOption match {
+          case Some(v) => // vertex frontier is not empty
+            // collect lambdas defined and library functions refered in this vertex
+            val lambdas = lambdaDefsIn(v.tree)
+            val libdefs = libdefRefsIn(v.tree)
+
+            // collect the call edges in this vertex
+            val edgBldr = Set.newBuilder[CG.Edge]
+            // helper method: adds edges representing function parameter bindings
+            val addBnds = (argss: Seq[Seq[u.Tree]], sym: u.MethodSymbol) =>
+              for {
+                (args, params) <- argss zip sym.paramLists
+                (TermRef(arg), TermSym(param)) <- args zip params
+                if lambdas.contains(arg)
+              } edgBldr += CG.Binds(CG.FunPar(param), lambdas(arg))
+            // main traversal that accumulates the edges in `edgBldr`
+            TopDown
+              .inherit({
+                // inherit enclosing caller lambda or library function
+                case ValDef(sym, Lambda(_, _, _), _) =>
+                  lambdas.getOrElse(sym, v)
+              })(Monoids.right(v))
+              .inherit({
+                // inherit function type parameters of enclosing Lambda and DefDef nodes
+                case Lambda(_, params, _) =>
+                  params.map(Sym.of).filter(s => Sym.funs(Type.of(s).typeSymbol)).toList
+                case DefDef(_, _, _, paramss, _) =>
+                  paramss.flatten.map(Sym.of).filter(s => Sym.funs(Type.of(s).typeSymbol)).toList
+              })
+              .traverseWith({
+                // function parameter call
+                case Attr.inh(DefCall(Some(TermRef(tgt)), sym, _, _@_*), enclFunPars :: caller :: _)
+                  if enclFunPars.contains(tgt) =>
+                  edgBldr += CG.Calls(caller, CG.FunPar(tgt))
```

Comment:
[Added `addBnds` in the third case as well](https://github.com/emmalanguage/emma/pull/253/commits/6a5390facaac77164f76939568254d99d5df76dc#diff-cd641dcce68f334f9299d35da1684e0cR314).